### Import Libraries

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('fifa21 data.csv')
pd.set_option('display.max_columns', None)
df

C:\Users\ASUS\AppData\Local\Temp\ipykernel_1300\887213838.py:1: DtypeWarning: Columns (76) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('fifa21 data.csv')


ID               Name                      LongName  \
0      158023           L. Messi                  Lionel Messi   
1       20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2      200389           J. Oblak                     Jan Oblak   
3      192985       K. De Bruyne               Kevin De Bruyne   
4      190871          Neymar Jr    Neymar da Silva Santos Jr.   
...       ...                ...                           ...   
18974  247223             Xia Ao                        Ao Xia   
18975  258760           B. Hough                     Ben Hough   
18976  252757        R. McKinley                Ronan McKinley   
18977  243790       Wang Zhen'ao                  Zhen'ao Wang   
18978  252520          Zhou Xiao                     Xiao Zhou   

                                               photoUrl  \
0      https://cdn.sofifa.com/players/158/023/21_60.png   
1      https://cdn.sofifa.com/players/020/801/21_60.png   
2      https://cdn.sofifa.com/players/200/389/21_60.png   
3      https://cdn.sofifa.com/players/192/985/21_60.png   
4      https://cdn.sofifa.com/players/190/871/21_60.png   
...                                                 ...   
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
0      http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1      http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
2      http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   
3      http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   
4      http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   
...                                                  ...         ...  ...   
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                         Club     Contract   Positions Height  \
0        93   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   
1        92   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   
2        91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   
3        91   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   
4        91   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   
...     ...  ...                          ...          ...         ...    ...   
18974    47   55           \n\n\n\nWuhan Zall  2018 ~ 2022          CB  178cm   
18975    47   67      \n\n\n\nOldham Athletic  2020 ~ 2021          CM  175cm   
18976    47   65           \n\n\n\nDerry City  2019 ~ 2020          CM  179cm   
18977    47   57     \n\n\n\nDalian YiFang FC  2020 ~ 2022          RW  175cm   
18978    47   57     \n\n\n\nDalian YiFang FC  2019 ~ 2023      CB, LB  188cm   

      Weight Preferred Foot  BOV Best Position     Joined Loan Date End  \
0       72kg           Left   93            RW  01-Jul-04           NaN   
1       83kg          Right   92            ST  10-Jul-18           NaN   
2       87kg          Right   91            GK  16-Jul-14           NaN   
3       70kg          Right   91           CAM  30-Aug-15           NaN   
4       68kg          Right   91            LW  03-Aug-17           NaN   
...      ...            ...  ...           ...        ...           ...   
18974   66kg          Right   49            CB  13-Jul-18           NaN   
18975   65k

In [3]:
df.shape

(18979, 77)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

# Data Cleaning

### Create a copy of DataFrame

In [5]:
data = df.copy()
data.head(5)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position     Joined Loan Date End    Value   Wage  \
0           Left   93            RW  01-Jul-04           NaN  €103.5M  €560K   
1          Right   92            ST  10-Jul-18           NaN     €63M  €220K   
2          Right   91            GK  16-Jul-14           NaN    €120M  €125K   
3          Right   91           CAM  30-Aug-15           NaN    €129M  €370K   
4          Right   91            LW  03-Aug-17           NaN    €132M  €270K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   
2        €159.4M         95        13         11                15   
3          €161M        407        94         82                55   
4        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalt

### 1. Club

Current for each player.

In [6]:
data['Club'].dtype

dtype('O')

In [7]:
data['Club'].unique()

array(['\n\n\n\nFC Barcelona', '\n\n\n\nJuventus',
       '\n\n\n\nAtlético Madrid', '\n\n\n\nManchester City',
       '\n\n\n\nParis Saint-Germain', '\n\n\n\nFC Bayern München',
       '\n\n\n\nLiverpool', '\n\n\n\nReal Madrid', '\n\n\n\nChelsea',
       '\n\n\n\nTottenham Hotspur', '\n\n\n\nInter', '\n\n\n\nNapoli',
       '\n\n\n\nBorussia Dortmund', '\n\n\n\nManchester United',
       '\n\n\n\nArsenal', '\n\n\n\nLazio', '\n\n\n\nLeicester City',
       '\n\n\n\nBorussia Mönchengladbach', '\n\n\n\nReal Sociedad',
       '\n\n\n\nAtalanta', '\n\n\n\nOlympique Lyonnais', '\n\n\n\nMilan',
       '\n\n\n\nVillarreal CF', '\n\n\n\nRB Leipzig', '\n\n\n\nCagliari',
       '\n\n\n\nAjax', '\n\n\n\nSL Benfica', '\n\n\n\nAS Monaco',
       '\n\n\n\nWolverhampton Wanderers', '\n\n\n\nEverton',
       '\n\n\n\nFiorentina', '\n\n\n\nFC Porto', '\n\n\n\nRC Celta',
       '\n\n\n\nTorino', '\n\n\n\nSevilla FC', '\n\n\n\nGrêmio',
       '\n\n\n\nReal Betis', '\n\n\n\nRoma', '\n\n\n\nNewcastle Unite

There are many whitespace characters in some club names in the Club column. Since the whitespace characters are both at the beginning and end of the string, using .strip() would be appropriate to remove them.

In [8]:
data['Club'] = data['Club'].str.strip()
data['Club'].unique()

array(['FC Barcelona', 'Juventus', 'Atlético Madrid', 'Manchester City',
       'Paris Saint-Germain', 'FC Bayern München', 'Liverpool',
       'Real Madrid', 'Chelsea', 'Tottenham Hotspur', 'Inter', 'Napoli',
       'Borussia Dortmund', 'Manchester United', 'Arsenal', 'Lazio',
       'Leicester City', 'Borussia Mönchengladbach', 'Real Sociedad',
       'Atalanta', 'Olympique Lyonnais', 'Milan', 'Villarreal CF',
       'RB Leipzig', 'Cagliari', 'Ajax', 'SL Benfica', 'AS Monaco',
       'Wolverhampton Wanderers', 'Everton', 'Fiorentina', 'FC Porto',
       'RC Celta', 'Torino', 'Sevilla FC', 'Grêmio', 'Real Betis', 'Roma',
       'Newcastle United', 'Eintracht Frankfurt', 'Valencia CF',
       'Medipol Başakşehir FK', 'Inter Miami', 'Bayer 04 Leverkusen',
       'Levante UD', 'Crystal Palace', 'Athletic Club de Bilbao',
       'Shanghai SIPG FC', 'VfL Wolfsburg',
       'Guangzhou Evergrande Taobao FC', 'Al Shabab',
       'Olympique de Marseille', 'Los Angeles FC',
       'Beijing Sino

### 2. Contract

In [9]:
data['Contract'].dtype

dtype('O')

In [10]:
data['Contract'].unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

using `.loc` to search for values in the 'Contract' column for each row where the value contains the string 'On Loan' or 'Free'.

In [11]:
for index, kontrak in data.loc[(data['Contract'].str.contains('On Loan')) | (data['Contract'].str.contains('Free')), 'Contract'].items():
    print(kontrak)

Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Free
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Free
Free
Free
Free
Free
Free
Free
Free
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Free
Free
Free
Jun 30, 2021 On Loan
Dec 31, 2020 On Loan
Jun 30, 2021 On Loan
Free
Jun 30, 2021 On Loan
Free
Free
Free
Free
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Jun 30, 2021 On Loan
Free
Jun 30, 2021 On Loan
Jan 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Dec 31, 2020 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Jun 30, 2021 On Loan
Free
Free
Free

Declare a function `extract_contract` that accepts one parameter `contract`, which is used to create 3 new columns: 'Contract Start', 'Contract End', and 'Contract Length (years)' with values taken from the 'Contract' column.

In [12]:
def extract_contract(contract):
    if contract == 'Free' or 'On Loan' in contract:
        start_date = np.nan
        end_date = np.nan
        contract_length = 0
    else:
        start_date, end_date = contract.split(' ~ ')
        start_year = int(start_date[:4])
        end_year = int(end_date[:4])
        contract_length = end_year - start_year
    return start_date, end_date, contract_length

new_columns = ['Contract Start', 'Contract End', 'Contract Length(years)']
new_data = data['Contract'].apply(lambda x: pd.Series(extract_contract(x)))

for i in range(len(new_columns)):
    data.insert(loc=data.columns.get_loc('Contract')+1+i, column=new_columns[i], value=new_data[i])

In [13]:
data[['Contract', 'Contract Start', 'Contract End', 'Contract Length(years)']].sample(10)

Contract Contract Start Contract End  \
10809           2020 ~ 2022           2020         2022   
8912            2019 ~ 2021           2019         2021   
8092   Jun 30, 2021 On Loan            NaN          NaN   
11143           2020 ~ 2024           2020         2024   
11312           2019 ~ 2021           2019         2021   
5625            2020 ~ 2023           2020         2023   
15620           2019 ~ 2024           2019         2024   
1848            2019 ~ 2024           2019         2024   
15556           2020 ~ 2021           2020         2021   
6223            2016 ~ 2022           2016         2022   

       Contract Length(years)  
10809                     2.0  
8912                      2.0  
8092                      0.0  
11143                     4.0  
11312                     2.0  
5625                      3.0  
15620                     5.0  
1848                      5.0  
15556                     1.0  
6223                      6.0

Declare a function `Categorize_Contract` that categorizes values in the 'Contract' column into three categories: 'Free', 'On Loan', and 'Contract'. Then create a new column 'Contract Status' to the DataFrame data to the right of the 'Contract Length (years)' column. The values for this new column are generated by applying the `Categorize_Contract` function to each value in the 'Contract' column.

In [14]:
def Categorize_Contract(contract):
    if contract == 'Free':
        return 'Free'
    elif 'On Loan'in contract:
        return 'On Loan'
    else:
        return 'Contract'

data.insert(data.columns.get_loc('Contract Length(years)')+1, 'Contract Status', data['Contract'].apply(Categorize_Contract))
data.sample(4)

ID          Name            LongName  \
12861  258271  Miguel Baeza  Miguel Baeza Pérez   
7322   176895      M. Pinto        Miguel Pinto   
11796  220872     G. Kiltie         Greg Kiltie   
1716   227174       M. Cash          Matty Cash   

                                               photoUrl  \
12861  https://cdn.sofifa.com/players/258/271/21_60.png   
7322   https://cdn.sofifa.com/players/176/895/21_60.png   
11796  https://cdn.sofifa.com/players/220/872/21_60.png   
1716   https://cdn.sofifa.com/players/227/174/21_60.png   

                                               playerUrl Nationality  Age  \
12861  http://sofifa.com/player/258271/miguel-baeza-p...       Spain   20   
7322   http://sofifa.com/player/176895/miguel-pinto/2...       Chile   36   
11796  http://sofifa.com/player/220872/greg-kiltie/21...    Scotland   23   
1716   http://sofifa.com/player/227174/matty-cash/210...     England   22   

       ↓OVA  POT         Club     Contract Contract Start Contract End  \
12861    63   72     RC Celta  2020 ~ 2025           2020         2025   
7322     68   68    Colo-Colo  2020 ~ 2020           2020         2020   
11796    64   71   Kilmarnock  2013 ~ 2021           2013         2021   
1716     75   82  Aston Villa  2020 ~ 2025           2020         2025   

       Contract Length(years) Contract Status    Positions Height Weight  \
12861                     5.0        Contract  CAM, LW, RW  177cm   67kg   
7322                      0.0        Contract           GK  183cm   79kg   
11796                     8.0        Contract      LM, CAM  173cm   70kg   
1716                      5.0        Contract       RB, RM  179cm   64kg   

      Preferred Foot  BOV Best Position     Joined Loan Date End  Value  Wage  \
12861           Left   65           CAM  15-Aug-20           NaN  €1.1M   €3K   
7322           Right   68            GK  08-Jan-20           NaN  €130K   €2K   
11796          Right   65           CAM  01-Jun-13           NaN  €1.2M   €2K   
1716           Right   75            RB  03-Sep-20           NaN   €11M  €38K   

      Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
12861          €1.3M        263        66         54                36   
7322           €210K        106        23         20                12   
11796          €1.2M        275        62         53                46   
1716          €17.9M        299        69         66                51   

       Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  \
12861             72       35    302         63     68           37   
7322              38       13    138         21     16           33   
11796             65       49    283         67     47           45   
1716              65       48    313         70     64           48   

       Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  \
12861            69            65       305            59            68   
7322             45            23       291            52            54   
11796            58            66       341            68            66   
1716             62            69       386            78            84   

       Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  \
12861       61         43       74    274          51       62       54   
7322        55         72       58    243          51       68       36   
11796       78         61       68    286          58       44       67   
1716        86         71       67    353          64       80       89   

       Strength  Long Shots  Mentality  Aggression  Interceptions  \
12861        48          59        241          49             26   
7322         68          20        154          27             11   
11796        65          52        272          55             42   
1716         56          64        300          56             72   

       Positioning  Vision  Penalties  Composure  Defending  Marking  \
12861           55

In [15]:
data[['Contract', 'Contract Start', 'Contract End', 'Contract Length(years)', 'Contract Status']].sample(5)

Contract Contract Start Contract End  Contract Length(years)  \
2385   2018 ~ 2022           2018         2022                     4.0   
15839  2020 ~ 2022           2020         2022                     2.0   
10817  2019 ~ 2023           2019         2023                     4.0   
9817   2005 ~ 2021           2005         2021                    16.0   
13271  2020 ~ 2023           2020         2023                     3.0   

      Contract Status  
2385         Contract  
15839        Contract  
10817        Contract  
9817         Contract  
13271        Contract

### 3. Height

In [16]:
data['Height'].dtype

dtype('O')

In [17]:
data['Height'].unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

Declare a function `transform_height` that accepts one parameter used to remove the string 'cm' and convert the height value from inches to centimeters, resulting integer data type only

In [18]:
def transform_height(x):
    if 'cm' in x:
        return int(x.strip('cm'))
    else:
        feet, inches = x.split("'")
        total_inches = int(feet)*12 + int(inches.strip('"'))
        return round(total_inches * 2.54)
    
data['Height'] = data['Height'].apply(transform_height)
data['Height'].unique()

array([170, 187, 188, 181, 175, 184, 191, 178, 193, 185, 199, 173, 168,
       176, 177, 183, 180, 189, 179, 195, 172, 182, 186, 192, 165, 194,
       167, 196, 163, 190, 174, 169, 171, 197, 200, 166, 164, 198, 201,
       158, 162, 161, 160, 203, 157, 156, 202, 159, 206, 155], dtype=int64)

In [19]:
data = data.rename(columns={'Height':'Height(Cm)'})
data.sample(1)

ID     Name     LongName  \
2679  207444  J. Damm  Jürgen Damm   

                                              photoUrl  \
2679  https://cdn.sofifa.com/players/207/444/21_60.png   

                                              playerUrl Nationality  Age  \
2679  http://sofifa.com/player/207444/jurgen-damm/21...      Mexico   27   

      ↓OVA  POT            Club     Contract Contract Start Contract End  \
2679    73   73  Atlanta United  2020 ~ 2021           2020         2021   

      Contract Length(years) Contract Status Positions  Height(Cm) Weight  \
2679                     1.0        Contract        RM         185   70kg   

     Preferred Foot  BOV Best Position     Joined Loan Date End  Value Wage  \
2679          Right   73            RM  12-Aug-20           NaN  €3.2M  €8K   

     Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
2679          €6.3M        318        67         65                65   

      Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  \
2679             70       51    331         73     66           59   

      Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  \
2679            61            72       405            92            92   

      Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  \
2679       88         70       63    336          65       73       74   

      Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  \
2679        69          55        307          68             48           70   

      Vision  Penalties  Composure  Defending  Marking  Standing Tackle  \
2679      60         61         64        118       42               39   

      Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  \
2679              37           60          8           15          16   

      GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  SM   A/W  \
2679              11           10         1875         407  3 ★  3★  High   

         D/W   IR  PAC  SHO  PAS  DRI  DEF  PHY Hits  
2679  Medium  3 ★   92   62   65   74   44   70   20

### 4. Weight

In [20]:
data['Weight'].dtype

dtype('O')

In [21]:
data['Weight'].unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

Declare a function `transform_weight` that removes the string 'kg' and 'lbs' characters, then converts the weight from pounds to kilograms, resulting integer data type

In [22]:
def transform_weight(x):
    if 'kg' in x:
        return int(x.strip('kg'))
    else:
        pounds = int(x.strip('lbs'))
        return round(pounds/2.205)
    
data['Weight'] = data['Weight'].apply(transform_weight)
data['Weight'].unique()

array([ 72,  83,  87,  70,  68,  80,  71,  91,  73,  85,  92,  69,  84,
        96,  81,  82,  75,  86,  89,  74,  76,  64,  78,  90,  66,  60,
        94,  79,  67,  65,  59,  61,  93,  88,  97,  77,  62,  63,  95,
       100,  58,  98, 103,  99, 102,  56, 101,  57,  55, 104, 107, 110,
        53,  50,  54,  52], dtype=int64)

In [23]:
data = data.rename(columns={'Weight':'Weight(kg)'})
data.sample(7)

ID            Name                      LongName  \
12499  230023         F. Luna                 Fernando Luna   
10126  213889       R. Haynes                   Ryan Haynes   
2127   195088     M. Ninković                Miloš Ninković   
13526  230290  Dêniro Prestes  Dêniro Rubens Prestes Cadete   
5770   214110     J. Cuadrado                 José Cuadrado   
16455  237208        P. Lewis                    Paul Lewis   
16098  254716         R. Báez                  Ronaldo Báez   

                                               photoUrl  \
12499  https://cdn.sofifa.com/players/230/023/21_60.png   
10126  https://cdn.sofifa.com/players/213/889/21_60.png   
2127   https://cdn.sofifa.com/players/195/088/21_60.png   
13526  https://cdn.sofifa.com/players/230/290/21_60.png   
5770   https://cdn.sofifa.com/players/214/110/21_60.png   
16455  https://cdn.sofifa.com/players/237/208/21_60.png   
16098  https://cdn.sofifa.com/players/254/716/21_60.png   

                                               playerUrl Nationality  Age  \
12499  http://sofifa.com/player/230023/fernando-luna/...   Argentina   30   
10126  http://sofifa.com/player/213889/ryan-haynes/21...     England   24   
2127   http://sofifa.com/player/195088/milos-ninkovic...      Serbia   35   
13526  http://sofifa.com/player/230290/deniro-rubens-...      Brazil   20   
5770   http://sofifa.com/player/214110/jose-cuadrado/...    Colombia   35   
16455  http://sofifa.com/player/237208/paul-lewis/210...     England   25   
16098  http://sofifa.com/player/254716/ronaldo-baez/2...    Paraguay   22   

       ↓OVA  POT                 Club     Contract Contract Start  \
12499    63   63            Patronato  2020 ~ 2021           2020   
10126    65   68       Newport County  2019 ~ 2021           2019   
2127     74   74            Sydney FC  2015 ~ 2022           2015   
13526    62   62  Ceará Sporting Club  2019 ~ 2023           2019   
5770     69   69    Atlético Nacional  2019 ~ 2020           2019   
16455    58   61      Tranmere Rovers  2020 ~ 2021           2020   
16098    59   69    Nacional Asunción  2019 ~ 2024           2019   

      Contract End  Contract Length(years) Contract Status    Positions  \
12499         2021                     1.0        Contract  CAM, LM, RM   
10126         2021                     2.0        Contract      LB, LWB   
2127          2022                     7.0        Contract      CAM, LM   
13526         2023                     4.0        Contract           ST   
5770          2020                     1.0        Contract           GK   
16455         2021                     1.0        Contract           CM   
16098         2024                     5.0        Contract           LM   

       Height(Cm)  Weight(kg) Preferred Foot  BOV Best Position     Joined  \
12499         175          70           Left   63           CAM  13-Jan-20   
10126         184          68           Left   66           LWB  12-Jul-19   
2127          180          76          Right   74           CAM  16-Jul-15   
13526         187          82          Right   62            ST  01-Jan-19   
5770          181          76          Right   69            GK  14-Jan-19   
16455         186          72          Right   60           CAM  17-Aug-20   
16098         168          69          Right   60            RM  01-Jul-19   

      Loan Date End  Value  Wage Release Clause  Attacking  Crossing  \
12499           NaN  €475K   €3K          €508K        253        41   
10126           NaN  €900K   €3K          €1.1M        272        63   
2127            NaN  €1.7M   €7K          €3.1M        331        69   
13526           NaN  €450K   €2K          €665K        271        33   
5770            NaN  €150K   €1K          €385K         90        13   
16455           NaN  €275K   €2K          €292K        237        45   
16098           NaN  €500K  €500          €588K        274        61   

       Finishing  Heading Accuracy  Short Passing  Volleys  

### 5. Loan Date End

Date when the player's loan ends (if On Loan)

In [24]:
data['Loan Date End'].dtype

dtype('O')

In [25]:
data['Loan Date End'].unique()

array([nan, '30-Jun-21', '31-Dec-20', '30-Jan-21', '30-Jun-22',
       '31-May-21', '05-Jul-21', '31-Dec-21', '01-Jul-21', '01-Jan-21',
       '31-Aug-21', '31-Jan-21', '30-Dec-21', '23-Jun-21', '03-Jan-21',
       '27-Nov-21', '17-Jan-21', '30-Jun-23', '31-Jul-21', '22-Nov-20',
       '31-May-22', '30-Dec-20', '04-Jan-21', '30-Nov-20', '01-Aug-21'],
      dtype=object)

In [26]:
player_onloan = data[data['Contract Status'] == 'On Loan']
player_onloan[['Contract','Contract Status', 'Contract End', 'Loan Date End']]

Contract Contract Status Contract End Loan Date End
205    Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
248    Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
254    Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
302    Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
306    Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
...                     ...             ...          ...           ...
18472  Aug 31, 2021 On Loan         On Loan          NaN     31-Aug-21
18571  Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
18600  Dec 31, 2020 On Loan         On Loan          NaN     31-Dec-20
18622  Dec 31, 2020 On Loan         On Loan          NaN     31-Dec-20
18680  Dec 31, 2020 On Loan         On Loan          NaN     31-Dec-20

[1013 rows x 4 columns]

There are many missing values in the `Loan Date End` column because not all players in this dataset are players in the "On Loan" status. There are 1013 rows of data from players in the "On Loan" status, and I think this column might be useful in the future, so I won't delete it.

### 6. W/F

Player's weak foot rating (out of 5)

In [27]:
data['W/F'].dtype

dtype('O')

In [28]:
data['W/F'].unique()

array(['4 ★', '3 ★', '5 ★', '2 ★', '1 ★'], dtype=object)

In [29]:
data['W/F'] = data['W/F'].str.replace('★', '')
data['W/F'].unique()

array(['4 ', '3 ', '5 ', '2 ', '1 '], dtype=object)

### 7. SM

Player's skill moves rating (out of 5)

In [30]:
data['SM'].dtype

dtype('O')

In [31]:
data['SM'].unique()

array(['4★', '5★', '1★', '2★', '3★'], dtype=object)

In [32]:
data['SM'] = data['SM'].str.replace('★', '')
data['SM'].unique()

array(['4', '5', '1', '2', '3'], dtype=object)

### 8. Hits

Number of times the player has been searched for in the FIFA database.

In [33]:
data['Hits'].dtype

dtype('O')

In [34]:
data['Hits'].unique()

array(['771', '562', '150', '207', '595', '248', '246', '120', '1.6K',
       '130', '321', '189', '175', '96', '118', '216', '212', '154',
       '205', '202', '339', '408', '103', '332', '86', '173', '161',
       '396', '1.1K', '433', '242', '206', '177', '1.5K', '198', '459',
       '117', '119', '209', '84', '187', '165', '203', '65', '336', '126',
       '313', '124', '145', '538', '182', '101', '45', '377', '99', '194',
       '403', '414', '593', '374', '245', '3.2K', '266', '299', '309',
       '215', '265', '211', '112', '337', '70', '159', '688', '116', '63',
       '144', '123', '71', '224', '113', '168', '61', '89', '137', '278',
       '75', '148', '176', '197', '264', '214', '247', '402', '440',
       '1.7K', '2.3K', '171', '320', '657', '87', '259', '200', '255',
       '253', '196', '60', '97', '85', '169', '256', '132', '239', '166',
       '121', '109', '32', '46', '122', '48', '527', '199', '282', '51',
       '1.9K', '642', '155', '323', '288', '497', '509', '79',

Declare a function `convert_hits` that replaces null values with 0, removes the string 'k' character, multiplies it by 1000, and converts it to an integer data type. Finally, if there's a float or integer value, it will be converted to an integer value only.

In [35]:
def convert_hits(x):
    if pd.isnull(x):
        return 0
    elif 'K' in str(x):
        number_only = str(x).replace('K', '')
        number_float = float(number_only)
        number_int = int(number_float * 1000)
        return number_int
    elif isinstance(x, float) or isinstance(x, int):
        return int(x)
    else:
        return int(float(x))

    
data['Hits'] = data['Hits'].apply(convert_hits)


In [36]:
data['Hits'].dtype

dtype('int64')

In [37]:
data['Hits'].unique()

array([ 771,  562,  150,  207,  595,  248,  246,  120, 1600,  130,  321,
        189,  175,   96,  118,  216,  212,  154,  205,  202,  339,  408,
        103,  332,   86,  173,  161,  396, 1100,  433,  242,  206,  177,
       1500,  198,  459,  117,  119,  209,   84,  187,  165,  203,   65,
        336,  126,  313,  124,  145,  538,  182,  101,   45,  377,   99,
        194,  403,  414,  593,  374,  245, 3200,  266,  299,  309,  215,
        265,  211,  112,  337,   70,  159,  688,  116,   63,  144,  123,
         71,  224,  113,  168,   61,   89,  137,  278,   75,  148,  176,
        197,  264,  214,  247,  402,  440, 1700, 2300,  171,  320,  657,
         87,  259,  200,  255,  253,  196,   60,   97,   85,  169,  256,
        132,  239,  166,  121,  109,   32,   46,  122,   48,  527,  199,
        282,   51, 1900,  642,  155,  323,  288,  497,  509,   79,   49,
        270,  511,   80,  128,  115,  156,  204,  143,  140,  152,  220,
        134,  225,   94,   74,  135,  142,   50,   

Since I want to use the 'Value', 'Wage', 'Release Clause' columns as aggregate columns in Power BI, it would be easier if I convert the values of these three columns into integers only value and add the euro denomination in the column names.

### 9. Value
Estimated market value of the player.

In [38]:
def convert_euro_to_number(value):
    if '€' in value:
        if 'M' in value:
            return int(float(value.replace('€', '').replace('M', '')) * 1e6)
        elif 'K' in value:
            return int(float(value.replace('€', '').replace('K', '')) * 1e3)
        else:
            return 0
    return int(value)

data['Value'] = data['Value'].apply(convert_euro_to_number)

In [39]:
data.rename(columns={'Value':'Value(Euro)'}, inplace=True)

In [40]:
data.sample(5)

ID             Name              LongName  \
7889   232826  F. Mac Allister  Francis Mac Allister   
15534  238070       J. Cumming         Jamie Cumming   
11466  229019         A. Engin           Ahmet Engin   
1775   176009          S. Kums             Sven Kums   
14924  247329        J. Khmiri         Jasser Khmiri   

                                               photoUrl  \
7889   https://cdn.sofifa.com/players/232/826/21_60.png   
15534  https://cdn.sofifa.com/players/238/070/21_60.png   
11466  https://cdn.sofifa.com/players/229/019/21_60.png   
1775   https://cdn.sofifa.com/players/176/009/21_60.png   
14924  https://cdn.sofifa.com/players/247/329/21_60.png   

                                               playerUrl Nationality  Age  \
7889   http://sofifa.com/player/232826/francis-mac-al...   Argentina   24   
15534  http://sofifa.com/player/238070/jamie-cumming/...     England   20   
11466  http://sofifa.com/player/229019/ahmet-engin/21...     Germany   23   
1775   http://sofifa.com/player/176009/sven-kums/210006/     Belgium   32   
14924  http://sofifa.com/player/247329/jasser-khmiri/...     Tunisia   22   

       ↓OVA  POT                    Club     Contract Contract Start  \
7889     67   75      Argentinos Juniors  2016 ~ 2023           2016   
15534    60   74               Stevenage  2017 ~ 2021           2017   
11466    64   70            MSV Duisburg  2015 ~ 2021           2015   
1775     75   75                KAA Gent  2020 ~ 2023           2020   
14924    60   70  Vancouver Whitecaps FC  2019 ~ 2023           2019   

      Contract End  Contract Length(years) Contract Status     Positions  \
7889          2023                     7.0        Contract       CDM, CM   
15534         2021                     4.0        Contract            GK   
11466         2021                     6.0        Contract        RM, LM   
1775          2023                     3.0        Contract  CM, CDM, CAM   
14924         2023                     4.0        Contract            CB   

       Height(Cm)  Weight(kg) Preferred Foot  BOV Best Position     Joined  \
7889          176          71          Right   68           CDM  01-Jan-16   
15534         186          79          Right   60            GK  16-Jan-17   
11466         180          70          Right   65            RM  01-Jul-15   
1775          175          73          Right   75            CM  01-Jul-20   
14924         192          85          Right   62            CB  01-Feb-19   

      Loan Date End  Value(Euro)  Wage Release Clause  Attacking  Crossing  \
7889            NaN      2100000   €4K          €1.7M        256        49   
15534           NaN       525000  €850          €891K         61        13   
11466           NaN       925000  €850            €1M        261        58   
1775            NaN      4400000  €22K          €7.5M        332        76   
14924           NaN       525000   €1K          €495K        192        34   

       Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  \
7889          35                58             70       44    293         67   
15534          5                10             26        7     67          9   
11466         54                41             60       48    272         66   
1775          57                52             80       67    384         73   
14924         19                63             48       28    181         36   

       Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
7889      45           45            66            70       329            69   
15534     12           14            20            12       185            20   
11466     46           52            44            64       373            78   
1775      76           79            78            78       335            61   
14924     25           29            48            43       266            56   

       Sprint Speed  Agility  Reactions  Balance  Power  Sho

In [41]:
data['Value(Euro)'].dtypes

dtype('int64')

### 10. Wage

Weekly wage of the player.

In [42]:
data['Wage'] = data['Wage'].apply(convert_euro_to_number)

In [43]:
data.rename(columns={'Wage':'Wage(Euro)'}, inplace=True)

In [44]:
data.sample(5)

ID         Name          LongName  \
7373   215570       G. Rea          Glen Rea   
1139   205362  M. Nastasić   Matija Nastasić   
488    189271  F. Coquelin  Francis Coquelin   
12927  204785   M. Mandron    Mikael Mandron   
12346  239951      G. Asís      Gonzalo Asís   

                                               photoUrl  \
7373   https://cdn.sofifa.com/players/215/570/21_60.png   
1139   https://cdn.sofifa.com/players/205/362/21_60.png   
488    https://cdn.sofifa.com/players/189/271/21_60.png   
12927  https://cdn.sofifa.com/players/204/785/21_60.png   
12346  https://cdn.sofifa.com/players/239/951/21_60.png   

                                               playerUrl          Nationality  \
7373    http://sofifa.com/player/215570/glen-rea/210006/  Republic of Ireland   
1139   http://sofifa.com/player/205362/matija-nastasi...               Serbia   
488    http://sofifa.com/player/189271/francis-coquel...               France   
12927  http://sofifa.com/player/204785/mikael-mandron...               France   
12346  http://sofifa.com/player/239951/gonzalo-asis/2...            Argentina   

       Age  ↓OVA  POT             Club     Contract Contract Start  \
7373    25    67   71       Luton Town  2016 ~ 2022           2016   
1139    27    77   79    FC Schalke 04  2015 ~ 2022           2015   
488     29    80   80    Villarreal CF  2020 ~ 2024           2020   
12927   25    63   67  Crewe Alexandra  2020 ~ 2021           2020   
12346   24    63   70    Independiente  2020 ~ 2021           2020   

      Contract End  Contract Length(years) Contract Status Positions  \
7373          2022                     6.0        Contract   CDM, CB   
1139          2022                     7.0        Contract        CB   
488           2024                     4.0        Contract   CM, CDM   
12927         2021                     1.0        Contract        ST   
12346         2021                     1.0        Contract        RB   

       Height(Cm)  Weight(kg) Preferred Foot  BOV Best Position     Joined  \
7373          184          73          Right   71            CB  01-Jul-16   
1139          188          83           Left   79            CB  14-Jan-15   
488           176          74          Right   81           CDM  12-Aug-20   
12927         191          82           Left   65            ST  01-Aug-20   
12346         174          65          Right   63            RB  12-Aug-20   

      Loan Date End  Value(Euro)  Wage(Euro) Release Clause  Attacking  \
7373            NaN      1300000        5000          €1.7M        233   
1139            NaN     11000000       26000         €16.9M        257   
488             NaN     20500000       41000         €29.4M        331   
12927           NaN       675000        3000          €867K        282   
12346           NaN       925000        3000          €765K        215   

       Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  \
7373         43         29                68             65       28    243   
1139         46         30                77             72       32    314   
488          69         64                63             82       53    349   
12927        47         60                64             59       52    266   
12346        51         36                51             42       35    237   

       Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  \
7373          57     32           28            61            65       309   
1139          43     72           68            68            63       339   
488           80     68           41            80            80       383   
12927         63     52           43            43            65       302   
12346         64     40           35            41            57       321   

       Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  \
7373             62            64       60         66       57    290   
1139             65   

In [45]:
data['Wage(Euro)'].dtypes

dtype('int64')

### 11. Release Clause

Release clause in the player's contract

In [46]:
data['Release Clause'] = data['Release Clause'].apply(convert_euro_to_number)

In [47]:
data.rename(columns={'Release Clause':'Release Clause(Euro)'}, inplace=True)

In [48]:
data.sample(5)

ID          Name          LongName  \
15704  237381   B. Petersen   Brylon Petersen   
15734  248394    J. Corinus    Jérémy Corinus   
12631  221859   Lee Ju Yong       Ju Yong Lee   
18791  248719  W. Aouachria  Wassim Aouachria   
13869  247268    S. Higashi    Shunki Higashi   

                                               photoUrl  \
15704  https://cdn.sofifa.com/players/237/381/21_60.png   
15734  https://cdn.sofifa.com/players/248/394/21_60.png   
12631  https://cdn.sofifa.com/players/221/859/21_60.png   
18791  https://cdn.sofifa.com/players/248/719/21_60.png   
13869  https://cdn.sofifa.com/players/247/268/21_60.png   

                                               playerUrl     Nationality  Age  \
15704  http://sofifa.com/player/237381/brylon-peterse...    South Africa   24   
15734  http://sofifa.com/player/248394/jeremy-corinus...          France   23   
12631  http://sofifa.com/player/221859/ju-yong-lee/21...  Korea Republic   27   
18791  http://sofifa.com/player/248719/wassim-aouachr...         Algeria   20   
13869  http://sofifa.com/player/247268/shunki-higashi...           Japan   19   

       ↓OVA  POT                    Club     Contract Contract Start  \
15704    59   64           Kaizer Chiefs  2016 ~ 2021           2016   
15734    59   67              AC Ajaccio  2019 ~ 2024           2019   
12631    63   63  Jeonbuk Hyundai Motors  2018 ~ 2025           2018   
18791    50   64       Charlton Athletic  2019 ~ 2021           2019   
13869    62   73     Sanfrecce Hiroshima  2018 ~ 2023           2018   

      Contract End  Contract Length(years) Contract Status Positions  \
15704         2021                     5.0        Contract        GK   
15734         2024                     5.0        Contract   CB, CDM   
12631         2025                     7.0        Contract        LB   
18791         2021                     2.0        Contract   CAM, ST   
13869         2023                     5.0        Contract        CF   

       Height(Cm)  Weight(kg) Preferred Foot  BOV Best Position     Joined  \
15704         188          79          Right   59            GK  22-Sep-16   
15734         187          79          Right   61            CB  31-Jan-19   
12631         180          78           Left   63            LB  13-Oct-18   
18791         188          75          Right   52           CAM  08-Jan-19   
13869         180          69           Left   65           CAM  02-Sep-18   

      Loan Date End  Value(Euro)  Wage(Euro)  Release Clause(Euro)  Attacking  \
15704           NaN       275000           0                364000         75   
15734           NaN       475000           0                410000        224   
12631           NaN       475000        2000                423000        288   
18791           NaN       110000        1000                143000        214   
13869           NaN       875000        1000                725000        288   

       Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  \
15704        13          8                14             29       11     87   
15734        46         28                59             59       32    241   
12631        63         60                51             58       56    279   
18791        46         39                36             53       40    235   
13869        61         59                58             62       48    301   

       Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  \
15704         13     14           14            26            20       157   
15734         52     41           39            55            54       287   
12631         62     53           50            52            62       357   
18791         48     44           38            53            52       274   
13869         65     55           56            58            67       318   

       Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  \
15704            20            36  

In [49]:
data['Release Clause(Euro)'].dtypes

dtype('int64')

### Look of all columns

In [50]:
data.sample(5)


ID             Name           LongName  \
14682  246230       M. Phaëton   Matthias Phaëton   
14167  225600      A. Al Ammar  Abdullah Al Ammar   
1172   220621      S. Benrahma      Saïd Benrahma   
4967   133249  E. Hallfreðsson  Emil Hallfreðsson   
12136  251686     D. Fernández   Damián Fernández   

                                               photoUrl  \
14682  https://cdn.sofifa.com/players/246/230/21_60.png   
14167  https://cdn.sofifa.com/players/225/600/21_60.png   
1172   https://cdn.sofifa.com/players/220/621/21_60.png   
4967   https://cdn.sofifa.com/players/133/249/21_60.png   
12136  https://cdn.sofifa.com/players/251/686/21_60.png   

                                               playerUrl   Nationality  Age  \
14682  http://sofifa.com/player/246230/matthias-phaet...        France   20   
14167  http://sofifa.com/player/225600/abdullah-al-am...  Saudi Arabia   26   
1172   http://sofifa.com/player/220621/said-benrahma/...       Algeria   24   
4967   http://sofifa.com/player/133249/emil-hallfredh...       Iceland   36   
12136  http://sofifa.com/player/251686/damian-fernand...     Argentina   19   

       ↓OVA  POT                  Club     Contract Contract Start  \
14682    61   74  En Avant de Guingamp  2018 ~ 2023           2018   
14167    61   63              Damac FC  2020 ~ 2025           2020   
1172     77   80             Brentford  2018 ~ 2022           2018   
4967     70   70               No Club         Free            NaN   
12136    63   75       Vélez Sarsfield  2019 ~ 2022           2019   

      Contract End  Contract Length(years) Contract Status Positions  \
14682         2023                     5.0        Contract        ST   
14167         2025                     5.0        Contract        LB   
1172          2022                     4.0        Contract        LW   
4967           NaN                     0.0            Free   CM, CDM   
12136         2022                     3.0        Contract        CB   

       Height(Cm)  Weight(kg) Preferred Foot  BOV Best Position     Joined  \
14682         185          75          Right   63            ST  01-Feb-18   
14167         177          70           Left   64            LM  29-Sep-20   
1172          172          67          Right   78           CAM  06-Jul-18   
4967          186          86           Left   71           CDM  06-Jan-20   
12136         180          74          Right   65            CB  03-Jul-19   

      Loan Date End  Value(Euro)  Wage(Euro)  Release Clause(Euro)  Attacking  \
14682           NaN       800000        1000               1200000        272   
14167           NaN       375000        3000                474000        275   
1172            NaN     14500000       43000              22700000        329   
4967            NaN            0           0                     0        317   
12136           NaN      1100000        2000               1000000        206   

       Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  \
14682        34         62                59             56       61    275   
14167        62         59                51             61       42    277   
1172         74         72                47             72       64    368   
4967         68         50                60             77       62    369   
12136        29         25                65             59       28    212   

       Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  \
14682         65     60           40            48            62       323   
14167         60     53           43            59            62       346   
1172          83     73           69            64            79       394   
4967          74     74           72            74            75       297   
12136         52     38           27            45            50       327   

       Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  \
14682            72            75  

In [51]:
data.to_csv('Fifa21_fix.csv', index=False)